docs/advanced_examples/KNearestNeighbors.ipynb

```diff
@@ -0,0 +1,431 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The novel aspect introduced by Concret-ML models is the hyperparameter `n_bits`, which represents\n",
+    "# the precision for quantizing input data\n",
+    "# This quantization step is essential after the training phase, since FHE exclusively operates\n",
+    "# over integers\n",
+    "\n",
+    "n_neighbors = 3\n",
+    "\n",
+    "concrete_knn = ConcreteKNeighborsClassifier(n_bits=3, n_neighbors=n_neighbors)\n",
+    "\n",
+    "# Fit both the Concrete ML and its equivalent float estimators on clear data\n",
+    "concrete_knn, sklearn_model = concrete_knn.fit_benchmark(X_train, y_train)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Compile the model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Saved in 47\n",
+      "Compilation time: 2.56 seconds\n"
+     ]
+    }
+   ],
+   "source": [
+    "# The compilation step aims to:\n",
+    "# - convert the quantized model to its FHE equivalent\n",
+    "# - create an executable operation graph\n",
+    "# - check the operation graph's compatibility with FHE\n",
+    "# - compute the maximum bit-width needed for model execution\n",
+    "# - determine cryptographic parameters necessary for generating secret keys and evaluation keys\n",
+    "time_begin = time.time()\n",
+    "circuit = concrete_knn.compile(X)\n",
+    "print(f\"Compilation time: {time.time() - time_begin:.2f} seconds\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Generating a key for an 8-bit circuit\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(f\"Generating a key for an {circuit.graph.maximum_integer_bit_width()}-bit circuit\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Key generation"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Key generation time: 829.81 seconds\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Note that this step may be time-consuming for circuits exceeding 8-bits\n",
+    "time_begin = time.time()\n",
+    "circuit.client.keygen(force=False)\n",
+    "print(f\"Key generation time: {time.time() - time_begin:.2f} seconds\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Inference with Concrete ML:\n",
+    "\n",
+    "a. __clear__: inference on unencrypted quantized data, without any FHE execution \n",
+    "\n",
+    "b. __Execution in FHE__: inference on encrypted data, using actual FHE execution\n",
```

Comment:
I would talk about simulation first, and then FHE exec, as we usually do, making the first two inferences the ones in clear !

Reply:
otherwise i would recalculate them.

Reply:
I'm not sure how the cell order is related to the markdown table. Also, top-k labels are done in the clear, so even if there is some kind of re-computation, we are talking of less than a second right ?

Reply:
In this notebook, we want to compare the topk label of cml and that of sckit learn.

For SK, the topk label is available in a function called kneigbours, which returns the distance and indices of the k nearest neighbors.

for cml, this function doesn't exist ( which makes me think I should add a warning for this function).
for cml, only topk labels are retrieved to avoid information leaks.

as i was doing the FHE inference on 1 example, self.topk_label had 1 example. 
that's why i changed the order, to get the simulated pridictions on the whole test set.

Nervermind, I've restored the original order,, it takes ~ 4 mn, and I'm computing the predictions for the whole test set, now

Reply:
> ( which makes me think I should add a warning for this function).

indeed you could add the method and raise a warning in `knn.py` , keep in mind that you'll need to update a test or add a new one to catch the error

> for cml, only topk labels are retrieved to avoid information leaks.

instead of a having a `_topk_labels` parameter (which you access here and is not great), would it be possible to have a similar/different `neighbors` method for CML models that only returns these labels ?